librad/src/git/local/url.rs

```diff
@@ -0,0 +1,107 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use thiserror::Error;
+
+use crate::{
+    hash::{self, Hash},
+    uri::{self, RadUrn},
+};
+
+#[derive(Clone)]
+pub struct LocalUrl {
+    repo: Hash,
+}
+
+impl LocalUrl {
+    pub fn repo(&self) -> &Hash {
+        &self.repo
+    }
+}
+
+impl Display for LocalUrl {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}://{}.git", super::URL_SCHEME, self.repo)
+    }
+}
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum ParseError {
+    #[error("Invalid scheme: {0}")]
+    InvalidScheme(String),
+
+    #[error("Cannot-be-a-base URL")]
+    CannotBeABase,
+
+    #[error("Malformed URL")]
+    Url(#[from] url::ParseError),
+
+    #[error(transparent)]
+    Hash(#[from] hash::ParseError),
+}
+
+impl FromStr for LocalUrl {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let url = url::Url::parse(s)?;
+        if url.scheme() != super::URL_SCHEME {
+            return Err(Self::Err::InvalidScheme(url.scheme().to_owned()));
+        }
+        if url.cannot_be_a_base() {
+            return Err(Self::Err::CannotBeABase);
+        }
+
+        let repo = url
+            .host_str()
+            .expect("we checked for cannot-be-a-base. qed")
```

Comment:
Might be safer if we do:
```rust
let repo = if url.cannot_be_base() {
  return Err(Self::Err::CannotBeABase);
} else {
  /* this logic */
};
```

That way it's explicit we checked and we're in the `else` branch. If someone comes to change this code later, they'll have to think why this was an `if/else`.

Reply:
Sure ya... I used to dislike this pattern of early returns in Rust, too, but then I learned that type inference often fails when you do `let foo = { /* ... */ }?;`, which I find even more annoying. Feel free to change.